manager/allocator/networkallocator/networkallocator.go

```diff
@@ -303,38 +324,84 @@ func OnInit(options *ServiceAllocationOpts) {
 // IsServiceAllocated returns if the passed service has its network resources allocated or not.
 // init bool indicates if the func is called during allocator initialization stage.
 func (na *NetworkAllocator) IsServiceAllocated(s *api.Service, flags ...func(*ServiceAllocationOpts)) bool {
-	var options ServiceAllocationOpts
+	return !na.IsServiceToBeAllocated(s, flags...)
+}
 
+// IsServiceToBeAllocated returns true if the passed service needs to have network resources allocated/updated.
+func (na *NetworkAllocator) IsServiceToBeAllocated(s *api.Service, flags ...func(*ServiceAllocationOpts)) bool {
+	var options ServiceAllocationOpts
 	for _, flag := range flags {
 		flag(&options)
 	}
 
+	// Always prefer NetworkAttachmentConfig in the TaskSpec
+	specNetworks := s.Spec.Task.Networks
+	if len(specNetworks) == 0 && len(s.Spec.Networks) != 0 {
+		specNetworks = s.Spec.Networks
+	}
+
 	// If endpoint mode is VIP and allocator does not have the
 	// service in VIP allocated set then it is not allocated.
-	if (len(s.Spec.Task.Networks) != 0 || len(s.Spec.Networks) != 0) &&
+	if (len(specNetworks) != 0) &&
 		(s.Spec.Endpoint == nil ||
 			s.Spec.Endpoint.Mode == api.ResolutionModeVirtualIP) {
 		if _, ok := na.services[s.ID]; !ok {
-			return false
+			return true
+		}
+
+		if s.Endpoint.VirtualIPs == nil {
```

Comment:
Do we also need to check if `s.Endpoint` is nil? There is a check for that below, but if it's really possible for `s.Endpoint` to be nil, this would panic.